engine/src/main/scala/cromwell/engine/io/IoActor.scala

```diff
@@ -0,0 +1,148 @@
+      case gcsBatch: GcsBatchCommandContext[_, _] => 0
+      case other => 1
+    }))
+    
+    // Sub flow for batched gcs requests
+    val batches = batchPartitioner.out(0) collect { case batch: GcsBatchCommandContext[_, _] => batch }
+    
+    // Sub flow for single nio requests
+    val defaults = batchPartitioner.out(1) collect { case default: DefaultCommandContext[_] => default }
+    
+    // Merge results from both flows back together
+    val merger = builder.add(Merge[IoResult](2))
+    
+    // Flow processing nio requests
+    val defaultFlowPorts = builder.add(defaultFlow)
+    
+    // Flow processing gcs batch requests
+    val batchFlowPorts = builder.add(gcsBatchFlow)
+
+    input ~> batchPartitioner
+             defaults.outlet ~> defaultFlowPorts ~> merger
+             batches.outlet ~> batchFlowPorts ~> merger
+    
+    FlowShape[IoCommandContext[_], IoResult](input.in, merger.out)
+  }
+
+  protected val throttledFlow = throttle map { t => 
+    Flow[IoCommandContext[_]]
+      .throttle(t.elements, t.per, t.maximumBurst, ThrottleMode.Shaping)
+      .via(flow)
+  } getOrElse flow
+  
+  override protected lazy val streamSource = source.via(throttledFlow)
+  
+  override def actorReceive: Receive = {
+    /* GCS Batch command with context */
+    case (clientContext: Any, gcsBatchCommand: GcsBatchIoCommand[_, _]) =>
+      val replyTo = sender()
+      val commandContext= GcsBatchCommandContext(gcsBatchCommand, replyTo, Option(clientContext))
+      sendToStream(commandContext)
+
+    /* GCS Batch command without context */
+    case gcsBatchCommand: GcsBatchIoCommand[_, _] =>
+      val replyTo = sender()
+      val commandContext= GcsBatchCommandContext(gcsBatchCommand, replyTo)
+      sendToStream(commandContext)
+
+    /* Default command with context */
+    case (clientContext: Any, command: IoCommand[_]) =>
+      val replyTo = sender()
+      val commandContext= DefaultCommandContext(command, replyTo, Option(clientContext))
+      sendToStream(commandContext)
+      
+    /* Default command without context */
+    case command: IoCommand[_] => 
+      val replyTo = sender()
+      val commandContext= DefaultCommandContext(command, replyTo)
+      sendToStream(commandContext)
+  }
+}
+
+trait IoCommandContext[T] extends StreamContext {
+  def request: IoCommand[T]
+  def replyTo: ActorRef
+  def fail(failure: Throwable): IoResult = (request.fail(failure), this)
+  def success(value: T): IoResult = (request.success(value), this)
+}
+
+object IoActor {
+  /** Flow that can consume an IoCommandContext and produce an IoResult */
+  type IoFlow = Flow[IoCommandContext[_], IoResult, NotUsed]
+  
+  /** Result type of an IoFlow, contains the original command context and the final IoAck response. */
+  type IoResult = (IoAck[_], IoCommandContext[_])
+  
+  /** Maximum number of times a command will be attempted: First attempt + 3 retries */
+  val MaxAttemptsNumber = 1 + 3
```

Comment:
why not just call this 4 w/ the explanation? I see why you did it this way but all it did was leave me starting off more ??? than I would have been w/ 4 being the magic number.